In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
emp_df = pd.read_csv('employee_data.csv')
insurance_df = pd.read_csv('insurance_data.csv')
vendor_df = pd.read_csv('vendor_data.csv')

In [3]:
emp_df.head()

AGENT_ID      AGENT_NAME DATE_OF_JOINING          ADDRESS_LINE1  \
0  AGENT00001       Ray Johns      1993-06-05       1402 Maggies Way   
1  AGENT00002   Angelo Borjon      2005-12-27         414 Tanya Pass   
2  AGENT00003  Candy Spellman      2003-09-02    606 National Street   
3  AGENT00004      Mary Smith      2004-09-23  235 Hugh Thomas Drive   
4  AGENT00005    Mildred Diaz      2011-06-21  3426 Broadview Street   

  ADDRESS_LINE2              CITY STATE  POSTAL_CODE  EMP_ROUTING_NUMBER  \
0           NaN  Waterbury Center    VT         5677            34584958   
1           NaN       Panama City    FL        32404           107363763   
2          #306      Fayetteville    AR        72701            81744097   
3           NaN       Panama City    FL        32404            67563771   
4           NaN        Montgomery    AL        36110           114951317   

      EMP_ACCT_NUMBER  
0  HKUN51252328472585  
1  OPIS19290040088204  
2  YSCJ67489688482590  
3  ZANG21285355574581  
4  DZFS82244494451134

In [4]:
emp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   AGENT_ID            1200 non-null   object
 1   AGENT_NAME          1200 non-null   object
 2   DATE_OF_JOINING     1200 non-null   object
 3   ADDRESS_LINE1       1200 non-null   object
 4   ADDRESS_LINE2       179 non-null    object
 5   CITY                1193 non-null   object
 6   STATE               1200 non-null   object
 7   POSTAL_CODE         1200 non-null   int64 
 8   EMP_ROUTING_NUMBER  1200 non-null   int64 
 9   EMP_ACCT_NUMBER     1200 non-null   object
dtypes: int64(2), object(8)
memory usage: 93.9+ KB


In [5]:
insurance_df.head(2)

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594    2018-04-21   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  ...  \
0  2020-05-16  2020-05-21         Health          157.13          9000  ...   
1  2020-05-13  2020-05-18       Property          141.71         26000  ...   

  CLAIM_STATUS INCIDENT_SEVERITY AUTHORITY_CONTACTED ANY_INJURY  \
0            A        Major Loss              Police          0   
1            A        Total Loss           Ambulance          1   

  POLICE_REPORT_AVAILABLE  INCIDENT_STATE INCIDENT_CITY  \
0                       1              GA      Savannah   
1                       0              AL    Montgomery   

  INCIDENT_HOUR_OF_THE_DAY    AGENT_ID  VENDOR_ID  
0                        4  AGENT00413  VNDR00556  
1                        0  AGENT00769  VNDR00592  

[2 rows x 38 columns]

In [6]:
insurance_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   TXN_DATE_TIME             10000 non-null  object 
 1   TRANSACTION_ID            10000 non-null  object 
 2   CUSTOMER_ID               10000 non-null  object 
 3   POLICY_NUMBER             10000 non-null  object 
 4   POLICY_EFF_DT             10000 non-null  object 
 5   LOSS_DT                   10000 non-null  object 
 6   REPORT_DT                 10000 non-null  object 
 7   INSURANCE_TYPE            10000 non-null  object 
 8   PREMIUM_AMOUNT            10000 non-null  float64
 9   CLAIM_AMOUNT              10000 non-null  int64  
 10  CUSTOMER_NAME             10000 non-null  object 
 11  ADDRESS_LINE1             10000 non-null  object 
 12  ADDRESS_LINE2             1495 non-null   object 
 13  CITY                      9946 non-null   object 
 14  STATE  

In [7]:
insurance_df['CLAIM_STATUS'].value_counts()

A    9497
D     503
Name: CLAIM_STATUS, dtype: int64

In [8]:
vendor_df.head()

VENDOR_ID              VENDOR_NAME                   ADDRESS_LINE1  \
0  VNDR00001  King, Proctor and Jones        2027 North Shannon Drive   
1  VNDR00002               Garcia Ltd          5701 East Shirley Lane   
2  VNDR00003               Cherry LLC            1217 Cottondale Road   
3  VNDR00004              Mays-Benson  227 West Montgomery Cross Road   
4  VNDR00005               Wilson PLC            23 North Hill Street   

  ADDRESS_LINE2          CITY STATE  POSTAL_CODE  
0            #5  Fayetteville    AR        72703  
1           NaN    Montgomery    AL        36117  
2           NaN    Montgomery    AL        36109  
3          #736      Savannah    GA        31406  
4           NaN     Nashville    TN        37210

In [9]:
print(emp_df.shape)
print(insurance_df.shape)
print(vendor_df.shape)

(1200, 10)
(10000, 38)
(600, 7)


## Task 1

**Merge the 3 dataset and create 1 view of data.
You can merge insurance_data.csv and employee_data.csv on
AGENT_ID You can merge insurance_data.csv and vendor_data.csv on
VENDOR_ID Note: Use left Outer join as not all claims require Vendor**

In [10]:
insurance_emp = pd.merge(insurance_df,emp_df, on ='AGENT_ID')

In [11]:
insurance_emp.head()

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-10 00:00:00    TXN00000258   A00003072   PLC00008557    2014-02-15   
2  2020-09-07 00:00:00    TXN00002428   A00001447   PLC00004391    2015-01-30   
3  2020-09-22 00:00:00    TXN00002811   A00001822   PLC00009369    2012-08-22   
4  2020-09-24 00:00:00    TXN00002861   A00006019   PLC00006747    2017-02-05   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  ...  \
0  2020-05-16  2020-05-21         Health          157.13          9000  ...   
1  2020-05-27  2020-06-01           Life           57.15         54000  ...   
2  2020-08-23  2020-08-28           Life           65.06         13000  ...   
3  2020-09-06  2020-09-11           Life           87.16         42000  ...   
4  2020-09-14  2020-09-17         Mobile           11.85           500  ...   

   VENDOR_ID   AGENT_NAME DATE_OF_JOINING      ADDRESS_LINE1_y  \
0  VNDR00556  Amy Wangler      1994-01-06  584 Rural Hill Road   
1        NaN  Amy Wangler      1994-01-06  584 Rural Hill Road   
2        NaN  Amy Wangler      1994-01-06  584 Rural Hill Road   
3        NaN  Amy Wangler      1994-01-06  584 Rural Hill Road   
4  VNDR00344  Amy Wangler      1994-01-06  584 Rural Hill Road   

  ADDRESS_LINE2_y     CITY_y STATE_y POSTAL_CODE_y  EMP_ROUTING_NUMBER  \
0             NaN  Nashville      TN         37217            35441155   
1             NaN  Nashville      TN         37217            35441155   
2             NaN  Nashville      TN         37217            35441155   
3             NaN  Nashville      TN         37217            35441155   
4             NaN  Nashville      TN         37217            35441155   

      EMP_ACCT_NUMBER  
0  LRZO53254369804934  
1  LRZO53254369804934  
2  LRZO53254369804934  
3  LRZO53254369804934  
4  LRZO53254369804934  

[5 rows x 47 columns]

In [12]:
insurance_ven = pd.merge(insurance_df,vendor_df,on='VENDOR_ID',how = 'left')

In [13]:
insurance_ven.head()

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594    2018-04-21   
2  2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969    2019-10-03   
3  2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292    2016-11-29   
4  2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204    2011-12-26   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  ...  \
0  2020-05-16  2020-05-21         Health          157.13          9000  ...   
1  2020-05-13  2020-05-18       Property          141.71         26000  ...   
2  2020-05-21  2020-05-26       Property          157.24         13000  ...   
3  2020-05-14  2020-05-19         Health          172.87         16000  ...   
4  2020-05-17  2020-05-22         Travel           88.53          3000  ...   

    INCIDENT_CITY INCIDENT_HOUR_OF_THE_DAY    AGENT_ID  VENDOR_ID  \
0        Savannah                        4  AGENT00413  VNDR00556   
1      Montgomery                        0  AGENT00769  VNDR00592   
2  Grand Junction                       19  AGENT00883  VNDR00031   
3        Savannah                       12  AGENT00278  VNDR00075   
4       Nashville                       18  AGENT00636  VNDR00472   

                        VENDOR_NAME            ADDRESS_LINE1_y  \
0                        Ramsey Inc       3114 West 20th Court   
1  Morales, Campbell and Washington      42 West Louise Street   
2                       Johnson Ltd    2619 North Quality Lane   
3                  Blackwell-Foster          3609 Meadow Drive   
4                       Evans-Scott  4001 19th Place Northeast   

  ADDRESS_LINE2_y        CITY_y  STATE_y  POSTAL_CODE_y  
0             NaN   Panama City       FL        32405.0  
1               B  Fayetteville       AR        72701.0  
2            #315  Fayetteville       AR        72703.0  
3             NaN     Nashville       TN        37215.0  
4             NaN    Washington       DC        20018.0  

[5 rows x 44 columns]

In [14]:
view = pd.merge(insurance_emp,insurance_ven,how='left')

In [15]:
view

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER  \
0     2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468   
1     2020-06-10 00:00:00    TXN00000258   A00003072   PLC00008557   
2     2020-09-07 00:00:00    TXN00002428   A00001447   PLC00004391   
3     2020-09-22 00:00:00    TXN00002811   A00001822   PLC00009369   
4     2020-09-24 00:00:00    TXN00002861   A00006019   PLC00006747   
...                   ...            ...         ...           ...   
9995  2021-04-28 00:00:00    TXN00008381   A00006144   PLC00005140   
9996  2021-05-02 00:00:00    TXN00008462   A00004250   PLC00009507   
9997  2021-06-08 00:00:00    TXN00009410   A00004412   PLC00005536   
9998  2021-06-21 00:00:00    TXN00009756   A00003267   PLC00007273   
9999  2021-05-11 00:00:00    TXN00008671   A00006937   PLC00007149   

     POLICY_EFF_DT     LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  \
0       2015-06-23  2020-05-16  2020-05-21         Health          157.13   
1       2014-02-15  2020-05-27  2020-06-01           Life           57.15   
2       2015-01-30  2020-08-23  2020-08-28           Life           65.06   
3       2012-08-22  2020-09-06  2020-09-11           Life           87.16   
4       2017-02-05  2020-09-14  2020-09-17         Mobile           11.85   
...            ...         ...         ...            ...             ...   
9995    2012-10-23  2021-04-15  2021-04-20           Life           64.16   
9996    2019-08-30  2021-04-20  2021-04-25         Mobile            9.76   
9997    2019-11-25  2021-06-05  2021-06-05          Motor           97.78   
9998    2012-09-29  2021-06-01  2021-06-06         Mobile            7.64   
9999    2016-01-11  2021-05-08  2021-05-08           Life           73.51   

      CLAIM_AMOUNT  ...       AGENT_NAME DATE_OF_JOINING      ADDRESS_LINE1_y  \
0             9000  ...      Amy Wangler      1994-01-06  584 Rural Hill Road   
1            54000  ...      Amy Wangler      1994-01-06  584 Rural Hill Road   
2            13000  ...      Amy Wangler      1994-01-06  584 Rural Hill Road   
3            42000  ...      Amy Wangler      1994-01-06  584 Rural Hill Road   
4              500  ...      Amy Wangler      1994-01-06  584 Rural Hill Road   
...            ...  ...              ...             ...                  ...   
9995         93000  ...  Angel Alexander      2012-08-26      44 Downey Drive   
9996           300  ...  Angel Alexander      2012-08-26      44 Downey Drive   
9997          3000  ...  Angel Alexander      2012-08-26      44 Downey Drive   
9998           200  ...  Angel Alexander      2012-08-26      44 Downey Drive   
9999         49000  ...  Charlotte Payne      2012-10-12     200 Boulder Road   

     ADDRESS_LINE2_y      CITY_y  STATE_y POSTAL_CODE_y EMP_ROUTING_NUMBER  \
0                NaN   Nashville       TN         37217           35441155   
1                NaN   Nashville       TN         37217           35441155   
2                NaN   Nashville       TN         37217           35441155   
3                NaN   Nashville       TN         37217           35441155   
4                NaN   Nashville       TN         37217           35441155   
...              ...         ...      ...           ...                ...   
9995             NaN  Manchester       CT          6040           55484383   
9996             NaN  Manchester       CT          6040           55484383   
9997             NaN  Manchester       CT          6040           55484383   
9998             NaN  Manchester       CT          6040           55484383   
9999             NaN  Manchester       CT          6040          118343345   

         EMP_ACCT_NUMBER  VENDOR_NAME  
0     LRZO53254369804934          NaN  
1     LRZO53254369804934          NaN  
2     LRZO53254369804934          NaN  
3     LRZO53254369804934          NaN  
4     LRZO53254369804934          NaN  
...                  ...          ...  
9995  CLKD05622028357384          NaN  
9996  CLKD056220

In [16]:
view.shape

(10000, 48)

In [17]:
view.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 0 to 9999
Data columns (total 48 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   TXN_DATE_TIME             10000 non-null  object 
 1   TRANSACTION_ID            10000 non-null  object 
 2   CUSTOMER_ID               10000 non-null  object 
 3   POLICY_NUMBER             10000 non-null  object 
 4   POLICY_EFF_DT             10000 non-null  object 
 5   LOSS_DT                   10000 non-null  object 
 6   REPORT_DT                 10000 non-null  object 
 7   INSURANCE_TYPE            10000 non-null  object 
 8   PREMIUM_AMOUNT            10000 non-null  float64
 9   CLAIM_AMOUNT              10000 non-null  int64  
 10  CUSTOMER_NAME             10000 non-null  object 
 11  ADDRESS_LINE1_x           10000 non-null  object 
 12  ADDRESS_LINE2_x           1495 non-null   object 
 13  CITY_x                    9946 non-null   object 
 14  STATE_x

## Task 2

**Business Leader wants to find Top 3 Insurance Type where we are getting
most insurance claims?**


In [20]:
view[['INSURANCE_TYPE','CLAIM_AMOUNT']].sort_values(by='CLAIM_AMOUNT', ascending=False).head(3)

INSURANCE_TYPE  CLAIM_AMOUNT
2828           Life        100000
4854           Life        100000
9120           Life        100000

## Task 3

**Business Leader wants to find Top 5 States where we are getting most insurance claims
for customer belonging to HIGH(H) risk segment?**

In [126]:
insurance_df[insurance_df['RISK_SEGMENTATION']=='H'].STATE.value_counts()[:5]

CA    148
AZ    105
FL    104
TN    100
AR     97
Name: STATE, dtype: int64

## Task 4
**Business wants to create a new variable “COLOCATION” which will have following values IF
Customer State == Incident State == Agent Address State THEN 1 ELSE 0
Find the mean of this new column**

In [21]:
view['COLOCATION'] = np.where((view['STATE_x'] == view['INCIDENT_STATE']) & (
    view['STATE_x'] == view['STATE_y']), 1, 0)
 
# printing the dataframe
view.head()

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-10 00:00:00    TXN00000258   A00003072   PLC00008557    2014-02-15   
2  2020-09-07 00:00:00    TXN00002428   A00001447   PLC00004391    2015-01-30   
3  2020-09-22 00:00:00    TXN00002811   A00001822   PLC00009369    2012-08-22   
4  2020-09-24 00:00:00    TXN00002861   A00006019   PLC00006747    2017-02-05   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  ...  \
0  2020-05-16  2020-05-21         Health          157.13          9000  ...   
1  2020-05-27  2020-06-01           Life           57.15         54000  ...   
2  2020-08-23  2020-08-28           Life           65.06         13000  ...   
3  2020-09-06  2020-09-11           Life           87.16         42000  ...   
4  2020-09-14  2020-09-17         Mobile           11.85           500  ...   

  DATE_OF_JOINING      ADDRESS_LINE1_y ADDRESS_LINE2_y     CITY_y STATE_y  \
0      1994-01-06  584 Rural Hill Road             NaN  Nashville      TN   
1      1994-01-06  584 Rural Hill Road             NaN  Nashville      TN   
2      1994-01-06  584 Rural Hill Road             NaN  Nashville      TN   
3      1994-01-06  584 Rural Hill Road             NaN  Nashville      TN   
4      1994-01-06  584 Rural Hill Road             NaN  Nashville      TN   

   POSTAL_CODE_y EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  VENDOR_NAME  \
0          37217           35441155  LRZO53254369804934          NaN   
1          37217           35441155  LRZO53254369804934          NaN   
2          37217           35441155  LRZO53254369804934          NaN   
3          37217           35441155  LRZO53254369804934          NaN   
4          37217           35441155  LRZO53254369804934          NaN   

   COLOCATION  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 49 columns]

In [24]:
view.COLOCATION.value_counts()

0    9956
1      44
Name: COLOCATION, dtype: int64

In [25]:
view.COLOCATION.mean()

0.0044

## Task 5

**Data entry error was detected in the data and you are required to correct it. If for any
claim transaction “AUTHORITY_CONTACTED” is NOT “Police” and POLICE_AVAILABLE == 1
Then Update “AUTHORITY_CONTACTED” to Police.**

In [29]:
view['POLICE_REPORT_AVAILABLE'] = np.where((view['AUTHORITY_CONTACTED'] == 'Police'), 1, 0)


In [30]:
(view[['AUTHORITY_CONTACTED','POLICE_REPORT_AVAILABLE']]).head(50)

AUTHORITY_CONTACTED  POLICE_REPORT_AVAILABLE
0               Police                        1
1                 None                        0
2            Ambulance                        0
3            Ambulance                        0
4                Other                        0
5               Police                        1
6            Ambulance                        0
7            Ambulance                        0
8               Police                        1
9                 None                        0
10           Ambulance                        0
11               Other                        0
12           Ambulance                        0
13           Ambulance                        0
14              Police                        1
15               Other                        0
16              Police                        1
17              Police                        1
18                None                        0
19              Police                        1
20              Police                        1
21           Ambulance                        0
22               Other                        0
23              Police                        1
24           Ambulance                        0
25           Ambulance                        0
26              Police                        1
27           Ambulance                        0
28           Ambulance                        0
29                None                        0
30                None                        0
31           Ambulance                        0
32              Police                        1
33           Ambulance                        0
34               Other                        0
35              Police                        1
36               Other                        0
37              Police                        1
38           Ambulance                        0
39              Police                        1
40           Ambulance                        0
41           Ambulance                        0
42              Police                        1
43           Ambulance                        0
44              Police                        1
45           Ambulance                        0
46           Ambulance                        0
47               Other                        0
48              Police                        1
49           Ambulance                        0

In [31]:
(view[['AUTHORITY_CONTACTED','POLICE_REPORT_AVAILABLE']].where(view.AUTHORITY_CONTACTED=='Police')).value_counts()

AUTHORITY_CONTACTED  POLICE_REPORT_AVAILABLE
Police               1.0                        2982
dtype: int64

## Task 6

**Business wants to check the Claim Amount for deviation for each transaction, they would
like you to calculate as follow
CLAIM_DEVIATION = AVG_CLAIM_AMOUNT_FOR_LAST_30DAYS (same insurance type)
/ CURRENT_CLAIM_AMOUNT
If the value < 0.5 THEN CLAIM_DEVIATION = 1 ELSE 0
If there is less than 30 days of transaction history THEN
-1 Note: LAST_30DAYS does not include current day**

In [93]:
view.TXN_DATE_TIME.max()

'2021-06-30 00:00:00'

In [94]:
# Assuming current day to be 2021-06-30
curr_date = pd.to_datetime('2021-06-30')
view['TXN_DATE_TIME'] = pd.to_datetime(view['TXN_DATE_TIME'])

In [95]:
# df with last 30 days data
last_30 = view.loc[(view['TXN_DATE_TIME'] > '2021-05-30') & (view['TXN_DATE_TIME'] < '2021-06-30')]  

# Getting average claim of last 30 days insurance type wise and storing it in dictionary
average_claim = last_30.groupby('INSURANCE_TYPE').mean()['CLAIM_AMOUNT'].to_dict()

# defining a function for calculating claim_deviation
def Claim_deviation(row):
    if (curr_date - row['TXN_DATE_TIME'])<pd.Timedelta(days=30):
        return -1
    claim_deviation = average_claim[row['INSURANCE_TYPE']] / row['CLAIM_AMOUNT']
    if claim_deviation < 0.5:
        return 1
    else:
        return 0
    

In [96]:
# Applying the function to the dataframe
view['CLAIM_DEVIATION'] = view.apply(Claim_deviation,axis=1)
view.CLAIM_DEVIATION.value_counts()

 0    9216
-1     784
Name: CLAIM_DEVIATION, dtype: int64

## Task 7

**Find All Agents who have worked on more than 2 types of Insurance Claims. Sort them by
Total Claim Amount Approved under them in descending order**

In [79]:
gb1 = (view[['AGENT_NAME','INSURANCE_TYPE', 'CLAIM_AMOUNT']]).groupby('AGENT_NAME').value_counts()

In [80]:
gb1

AGENT_NAME  INSURANCE_TYPE  CLAIM_AMOUNT
Adam Dixon  Health          8000            1
                            16000           1
            Life            44000           1
            Motor           5000            1
            Property        40000           1
                                           ..
Zella Sons  Mobile          200             1
                            300             1
                            400             1
            Property        40000           1
            Travel          1000            1
Length: 9449, dtype: int64

In [42]:
view.AGENT_NAME.value_counts()

Robert Ward         25
James Velasco       19
Frances Kimble      18
Larry Scott         17
Debbra Collins      17
                    ..
Sandra Tucker        2
Florence Ashlock     2
Nichole Wright       2
Harold Bonn          1
Charlotte Payne      1
Name: AGENT_NAME, Length: 1198, dtype: int64

In [43]:
view.INSURANCE_TYPE.value_counts()

Mobile      1692
Property    1692
Health      1690
Life        1682
Travel      1670
Motor       1574
Name: INSURANCE_TYPE, dtype: int64

## Task 8

**Mobile & Travel Insurance premium are discounted by 10%
Health and Property Insurance premium are increased by 7%
Life and Motor Insurance premium are marginally increased by 2%
What will be overall change in % of the Premium Amount Collected for all these Customer?**

In [97]:
premium_collected_earlier = view['PREMIUM_AMOUNT'].sum()

In [98]:
# total premium collected by each insurance type
prem_by_types = view.groupby('INSURANCE_TYPE').sum()['PREMIUM_AMOUNT'].to_dict()

In [99]:
prem_by_types

{'Health': 252455.73,
 'Life': 125621.22,
 'Mobile': 15191.42,
 'Motor': 165391.98,
 'Property': 202285.83,
 'Travel': 124139.77}

In [100]:
# Mobile & Travel premium discounted by 10%
prem_by_types['Mobile']*=0.9
prem_by_types['Travel']*=0.9
# Health and Property premium increases by 7%
prem_by_types['Health']*=1.07
prem_by_types['Property']*=1.07
# Life and Motor premium increased by 2%
prem_by_types['Life']*=1.02
prem_by_types['Motor']*=1.02

In [101]:
# premium after discount and increase
revised_premium = sum(prem_by_types.values())

In [102]:
(revised_premium-premium_collected_earlier)*100/premium_collected_earlier

2.6798588543858517

In [103]:
premium_collected_earlier,revised_premium

(885085.95, 908805.0042000001)

## Task 9

**Business wants to give discount to customer who are loyal and under stress due to Covid19.**
**They have laid down an eligibility Criteria as follow
IF CUSTOMER_TENURE > 60 AND EMPLOYMENT_STATUS = “N”
AND NO_OF_FAMILY_MEMBERS >=4 THEN 1 ELSE 0Create a new column “ELIGIBLE_FOR_DISCOUNT” and find it
mean.**

In [104]:
# calculating status for discount with given conditions
disc_status=(view['TENURE']>60)&(view['EMPLOYMENT_STATUS']=='N')&(view['NO_OF_FAMILY_MEMBERS']>=4)

In [105]:
# Creating an column ELIGIBLE_FOR_DISCOUNT with above values
view['ELIGIBLE_FOR_DISCOUNT'] = disc_status.astype('int')

In [106]:
# Mean of this column
view.ELIGIBLE_FOR_DISCOUNT.mean()

0.0299

## Task 10

**Business wants to check Claim Velocity which is defined as follow
CLAIM_VELOCITY = NO_OF_CLAIMS_IN_LAST30DAYS (for the current insurance type)
/ NO_OF_CLAIMS_IN_LAST3DAYS (for the current insurance type)
Note: LAST30DAYS & LAST3DAYS does not include current
day**

In [107]:
curr_date

Timestamp('2021-06-30 00:00:00')

In [108]:
view['CLAIM_STATUS'].value_counts()

A    9497
D     503
Name: CLAIM_STATUS, dtype: int64

In [109]:
last_3_days = view.loc[(view['TXN_DATE_TIME'] > curr_date-pd.Timedelta(days=4)) & (view['TXN_DATE_TIME'] < curr_date)]

In [110]:
# Calculating total approved claims insurance type wise in last 30 days
claims_in_last_30_days = pd.DataFrame.from_dict(last_30[last_30['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST30DAYS'])

In [111]:
# Calculating total approved claims insurance type wise in last 3 days
claims_in_last_3_days = pd.DataFrame.from_dict(last_3_days[last_3_days['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST3DAYS'])

In [112]:
approved_claims = claims_in_last_30_days.join(claims_in_last_3_days)

In [113]:
approved_claims['CLAIM_VELOCITY'] = approved_claims['NO_OF_CLAIMS_IN_LAST30DAYS']/approved_claims['NO_OF_CLAIMS_IN_LAST3DAYS']

In [114]:
approved_claims

NO_OF_CLAIMS_IN_LAST30DAYS  NO_OF_CLAIMS_IN_LAST3DAYS  \
Health                           128                         14   
Life                             128                         12   
Mobile                           127                          9   
Motor                            117                         12   
Property                         125                         12   
Travel                           123                         14   

          CLAIM_VELOCITY  
Health          9.142857  
Life           10.666667  
Mobile         14.111111  
Motor           9.750000  
Property       10.416667  
Travel          8.785714

In [115]:
All_claims30D = pd.DataFrame.from_dict(last_30.groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST30DAYS'])

In [116]:
All_claims3D = pd.DataFrame.from_dict(last_3_days.groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST3DAYS'])

In [117]:
total_claims=All_claims30D.join(All_claims3D)

In [118]:
total_claims['CLAIM_VELOCITY'] = total_claims['NO_OF_CLAIMS_IN_LAST30DAYS']/total_claims['NO_OF_CLAIMS_IN_LAST3DAYS']
total_claims

NO_OF_CLAIMS_IN_LAST30DAYS  NO_OF_CLAIMS_IN_LAST3DAYS  \
Health                           136                         14   
Life                             140                         14   
Mobile                           134                         10   
Motor                            122                         13   
Property                         129                         13   
Travel                           125                         14   

          CLAIM_VELOCITY  
Health          9.714286  
Life           10.000000  
Mobile         13.400000  
Motor           9.384615  
Property        9.923077  
Travel          8.928571

## Task 11

**Find all low performing agents i.e. employees who are in the bottom 5 percentile based
on Claims worked by them.**

In [119]:
bottom_5percentile_limit = view.groupby('AGENT_ID').sum().quantile(0.05)['CLAIM_AMOUNT']
Agent_grouped_data = view.groupby('AGENT_ID').sum()

# Low performing agent data

low_performing_agents = Agent_grouped_data[Agent_grouped_data['CLAIM_AMOUNT']<=bottom_5percentile_limit][['CLAIM_AMOUNT']]
low_performing_agents

CLAIM_AMOUNT
AGENT_ID                
AGENT00006         22400
AGENT00014          7400
AGENT00016         27000
AGENT00085         25700
AGENT00088         27700
AGENT00091         30000
AGENT00098         28200
AGENT00134         22000
AGENT00149         28900
AGENT00156         13200
AGENT00172         22200
AGENT00185         16100
AGENT00186         27700
AGENT00188         28400
AGENT00197         18500
AGENT00252         20000
AGENT00263         11500
AGENT00275         16200
AGENT00279         28000
AGENT00300         12600
AGENT00310         28800
AGENT00316         22200
AGENT00351         21700
AGENT00425         17800
AGENT00442         30300
AGENT00536         20000
AGENT00562         21100
AGENT00571         24200
AGENT00585         16600
AGENT00590         22200
AGENT00604          5500
AGENT00621          5300
AGENT00628          6100
AGENT00672         21300
AGENT00689         13700
AGENT00706         10800
AGENT00721          9000
AGENT00732         11300
AGENT00773         15000
AGENT00780         11300
AGENT00790          6000
AGENT00804         20000
AGENT00805         22600
AGENT00830         23500
AGENT00841         17800
AGENT00856         24200
AGENT00894         27600
AGENT00910         15500
AGENT00969         28300
AGENT00985         30000
AGENT01021         24300
AGENT01037          5600
AGENT01052         28000
AGENT01077          4200
AGENT01107         24700
AGENT01128         28000
AGENT01131         27500
AGENT01151         28000
AGENT01154           800
AGENT01179         27300

## Task 12

**Business wants to find all Suspicious Employees (Agents).
IF TOTAL CLAIM AMOUNT which meet below criteria is >= 15000 THEN AGENT is classified
as Suspicious ELSE Not
CLAIM_STATUS = Approved AND CUSTOMER_RISK_SEGMENTATION = High
AND INCIDENT_SEVERITY = “Major Loss”
If Suspicious, then 1 ELSE 0. Find mean of this column.**

In [120]:
criteria = (view['CLAIM_STATUS']=='A')&(view['RISK_SEGMENTATION']=='H')&(view['INCIDENT_SEVERITY']=='Major Loss')
# transactions that follow criteria
criteriamet = view[criteria]
grouped_suspicion_data = criteriamet.groupby('AGENT_ID').sum()

In [122]:
# storing agents_id that are suspicious(met all the criteria)
suspicious_agents = list(grouped_suspicion_data[grouped_suspicion_data['CLAIM_AMOUNT']>=15000].index)
emp_df

AGENT_ID         AGENT_NAME DATE_OF_JOINING  \
0     AGENT00001          Ray Johns      1993-06-05   
1     AGENT00002      Angelo Borjon      2005-12-27   
2     AGENT00003     Candy Spellman      2003-09-02   
3     AGENT00004         Mary Smith      2004-09-23   
4     AGENT00005       Mildred Diaz      2011-06-21   
...          ...                ...             ...   
1195  AGENT01196     Myrtle Mccatty      2005-09-14   
1196  AGENT01197        John Dorsey      1993-12-04   
1197  AGENT01198        Ruth Bigbee      1993-09-05   
1198  AGENT01199     Tommie Camacho      2004-10-25   
1199  AGENT01200  Virginia Stachura      1999-05-25   

                         ADDRESS_LINE1 ADDRESS_LINE2              CITY STATE  \
0                     1402 Maggies Way           NaN  Waterbury Center    VT   
1                       414 Tanya Pass           NaN       Panama City    FL   
2                  606 National Street          #306      Fayetteville    AR   
3                235 Hugh Thomas Drive           NaN       Panama City    FL   
4                3426 Broadview Street           NaN        Montgomery    AL   
...                                ...           ...               ...   ...   
1195  3553 Brandywine Street Northwest           NaN        Washington    DC   
1196              2315 Chestnut Street           NaN        Montgomery    AL   
1197            1983 Reidsville Street           NaN         Annapolis    MD   
1198                  5821 Taft Street           NaN            Arvada    CO   
1199                   3579 2nd Avenue           NaN         Edgewater    MD   

      POSTAL_CODE  EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  
0            5677            34584958  HKUN51252328472585  
1           32404           107363763  OPIS19290040088204  
2           72701            81744097  YSCJ67489688482590  
3           32404            67563771  ZANG21285355574581  
4           36110           114951317  DZFS82244494451134  
...           ...                 ...                 ...  
1195        20008            32137938  EWBR19907574832501  
1196        36107            70796687  KGEJ74846009867837  
1197        21401           124993912  RASV88064214164820  
1198        80004            19424059  UOFO11656245339211  
1199        21037            12226940  ALYY24945545778511  

[1200 rows x 10 columns]

In [123]:
emp_df['SUSPICIOUS'] = 0
sus_emp_slice = emp_df['AGENT_ID'].isin(suspicious_agents)
# Changing suspicious column value to 1 of suspicious employees
emp_df.loc[sus_emp_slice,'SUSPICIOUS']=1

In [125]:
emp_df['SUSPICIOUS'].mean()
0.12416666666666666


0.12416666666666666

The mean of suspicious column is 0.124167 that means from all employees 12.42% are suspicious